### PR TITLE
CAMEL-20889: camel-core: Stream is not reset when Message.getBody(cla…

### PR DESCRIPTION
**Pull request by luigidemasi**
…ss) is invoked and stream caching is enabled

### CODE REVIEW

**Review by orpiske — CHANGES_REQUESTED** [review]

File: core/camel-base/src/main/java/org/apache/camel/impl/converter/CoreTypeConverterRegistry.java [hunk 1]
```diff
@@ -96,6 +97,10 @@ private <T> T fastConvertTo(Class<T> type, Exchange exchange, Object value) {
         if (value == null) {
             return null;
         }
+        //Reset the stream before trying to convert it.
+        if (value instanceof StreamCache) {
+            ((StreamCache) value).reset();
+        }
```
[the remark this answers is not in the document]
Comment (davsclaus) [reply core/camel-base/src/main/java/org/apache/camel/impl/converter/CoreTypeConverterRegistry.java:103]: See my comments in JIRA that is a better place to improve this.
And the JDK has this "instanceof" performance problem.

**Comment by orpiske** [comment]
I'm not entirely sure what we have decided for this one (see the comments on the ticket), but in any case, this needs a full test. I am running it on my CI and I'll report the results here once they are complete.

**Comment by orpiske** [comment]
> I'm not entirely sure what we have decided for this one (see the comments on the ticket), but in any case, this needs a full test. I am running it on my CI and I'll report the results here once they are complete.

Hey @luigidemasi, @davsclaus suggested an elegant solution of how to implement this without incurring on performance impacts. Please, could you take a look at that and try an implementation using it?

File: core/camel-support/src/main/java/org/apache/camel/support/ExchangeHelper.java [hunk 2]
```diff
@@ -1248,4 +1249,22 @@ public static <T> T getVariable(Exchange exchange, String name, Class<T> type) {
...
+     * the body.
+     *
+     * @param  exchange the message exchange being processed
+     * @param  type     the type to convert to
+     * @return the body of the message as the specified type, or <tt>null</tt> if body does not
+     *         exist
+     */
+    public static <T> T getBodyAndResetStreamCache(Exchange exchange, Class<T> type) {
+        Object body = exchange.getMessage().getBody();
+        if ((body instanceof StreamCache)) {
```
Comment (davsclaus) [inline core/camel-support/src/main/java/org/apache/camel/support/ExchangeHelper.java:1264]: it seems you have double paranthesis

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
Can you format the code 

```
mvn net.revelc.code.formatter:formatter-maven-plugin:format
```

**Pull request merged**